cmake/Helpers.cmake

```diff
@@ -0,0 +1,685 @@
+        ENDIF()
+        SET_PATH( DOC_DIR ${CMAKE_INSTALL_PREFIX}/share/doc/gemrb )
+        SET_PATH( ICON_DIR ${CMAKE_INSTALL_PREFIX}/share/pixmaps )
+        SET_PATH( SVG_DIR ${CMAKE_INSTALL_PREFIX}/share/icons/hicolor/scalable/apps )
+        SET_PATH( MENU_DIR ${CMAKE_INSTALL_PREFIX}/share/applications )
+        SET_PATH( EXAMPLE_CONF_DIR ${SYSCONF_DIR} )
+        SET_PATH( METAINFO_DIR ${CMAKE_INSTALL_PREFIX}/share/metainfo )
+    ELSEIF (${LAYOUT} MATCHES "opt")
+        SET_PATH( LIB_DIR ${CMAKE_INSTALL_PREFIX}/lib${LIBDIR_SUFFIX} )
+        SET_PATH( PLUGIN_DIR ${LIB_DIR}/plugins )
+        SET_PATH( DATA_DIR ${CMAKE_INSTALL_PREFIX}/share/ )
+        SET_PATH( MAN_DIR ${CMAKE_INSTALL_PREFIX}/man/man6 )
+        SET_PATH( BIN_DIR ${CMAKE_INSTALL_PREFIX}/bin )
+        SET_PATH( SYSCONF_DIR ${CMAKE_INSTALL_PREFIX}/etc )
+        SET_PATH( DOC_DIR ${CMAKE_INSTALL_PREFIX}/share/doc/gemrb )
+        SET_PATH( ICON_DIR ${CMAKE_INSTALL_PREFIX}/share/pixmaps )
+        SET_PATH( SVG_DIR ${CMAKE_INSTALL_PREFIX}/share/icons/hicolor/scalable/apps )
+        SET_PATH( MENU_DIR ${CMAKE_INSTALL_PREFIX}/share/applications )
+        SET_PATH( EXAMPLE_CONF_DIR ${SYSCONF_DIR} )
+        SET_PATH( METAINFO_DIR ${CMAKE_INSTALL_PREFIX}/share/metainfo )
+    ELSE (${LAYOUT} MATCHES "bundle") # Mac or iOS
+        SET(CMAKE_INSTALL_RPATH @loader_path/../Frameworks)
+        SET(CMAKE_BUILD_WITH_INSTALL_RPATH TRUE)
+        SET_PATH( BIN_DIR /Applications )
+        SET_PATH( LIB_DIR @loader_path/../Frameworks )
+        SET_PATH( METAINFO_DIR "" )
+    ENDIF (${LAYOUT} MATCHES "home")
+ENDFUNCTION()
+
+FUNCTION(READ_GEMRB_VERSION)
+    # try to extract the version from the source
+    FILE(READ ${CMAKE_CURRENT_SOURCE_DIR}/gemrb/core/InterfaceConfig.h GLOBALS)
+    SET(GEMRB_VERSION "")
+    STRING(REGEX MATCH "define VERSION_GEMRB .([^\"]*)" GEMRB_VERSION "${GLOBALS}")
+    STRING(REGEX REPLACE "define VERSION_GEMRB .([^\"]*)$" "\\1" GEMRB_VERSION "${GEMRB_VERSION}")
+    IF(GEMRB_VERSION STREQUAL "") # lookup failed
+        set(GEMRB_VERSION "unknown")
+    ENDIF()
+    message(STATUS "Detected version: ${GEMRB_VERSION}")
+    SET_INTERNAL(GEMRB_VERSION ${GEMRB_VERSION})
+    unset(GLOBALS)
+ENDFUNCTION()
+
+FUNCTION(CACHE_BUILD_TYPE)
+    # If the user specifies -DCMAKE_BUILD_TYPE on the command line, take their definition
+    # and dump it in the cache along with proper documentation, otherwise set CMAKE_BUILD_TYPE
+    # to Release prior to calling PROJECT()
+    IF(DEFINED CMAKE_BUILD_TYPE)
+        SET(CMAKE_BUILD_TYPE ${CMAKE_BUILD_TYPE} CACHE STRING "Choose the type of build, options are: None(CMAKE_CXX_FLAGS or CMAKE_C_FLAGS used) Debug Release RelWithDebInfo MinSizeRel.")
+    ELSE(DEFINED CMAKE_BUILD_TYPE)
+        SET(CMAKE_BUILD_TYPE RelWithDebInfo CACHE STRING "Choose the type of build, options are: None(CMAKE_CXX_FLAGS or CMAKE_C_FLAGS used) Debug Release RelWithDebInfo MinSizeRel.")
+    ENDIF(DEFINED CMAKE_BUILD_TYPE)
```

Comment:
As a further improvement, this can also be deduped using the `$<IF>` construct.

Reply:
If I understand it correctly, cmake generator expressions are evaluated during the cmake build step, whereas variable are set earlier, at the configuration step, so this is not possible. Or maybe I don't get your idea. :D

Reply:
Ah, ok. We could still reduce the duplication by using a temporary var though.